Fix #18689: Exclude current skill and its prerequisites from untriaged skills in the skill selector

Zivanovic8 opened the pull request:
## Overview

1. This PR fixes or fixes part of #18689.
2. This PR does the following: 
- Added an object _skillIdsToExclude_ to track skills that should be excluded.
- _skillIdsToExclude_ includes the current skill and all its prerequisites.
- Added the filter in _searchInUntriagedSkillSummaries_ to exclude these skills from the _Untriaged Skills_.
3. (For bug-fixing PRs only) The original bug occurred because: There was no filtering mechanism in _searchInUntriagedSkillSummaries_ function.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

### Before


https://github.com/user-attachments/assets/9d9a3c61-a8b3-4952-820a-a263a31a1cbb



### After


https://github.com/user-attachments/assets/8725a3ba-16bb-4bd8-aa6f-d53219d836f9



## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

Zivanovic8 committed:
Fixed skill selector so that it does not display the current skill or its prerequisites in Untriaged Skills by filtering out already selected skills and that skill itself.

Zivanovic8 committed:
Merge branch 'develop' into fixSkillPrerequisite

Zivanovic8 commented:
@seanlip PTAL

himkar-singh committed:
Tracking metrics for diagnostic test (#22068)

* Diagnostic Test Feature

* Fix test case

* added changes

* Changes as requested

* Added test cases

* Updated tests cases

seanlip commented:
Assigning @amyyeung17 for shadow review. 

@amyyeung17 please feel free to assign me once you think this PR is ready for merge. Thanks!

himkar-singh committed:
Follow up PR for Tracking metrics for diagnostic test. (#22078)

name changed

galaxyhub9 committed:
Fix #22061: Can Upload Media in Skill Editor (#22071)

Fix:Added RteHelperService to Providers in skill-editor-page.module.ts

amyyeung17 commented on core/templates/components/skill-selector/skill-selector.component.ts:
```diff
@@ -47,6 +47,7 @@ export class SkillSelectorComponent implements OnInit {
   @Input() categorizedSkills!: CategorizedSkills;
   @Input() untriagedSkillSummaries!: SkillSummary[];
   @Input() allowSkillsFromOtherTopics!: boolean;
+  @Input() skillIdsToExclude!: {[key: string]: boolean};
```
If we're only checking membership and each skill id is unique, I recommend using a Set. Sets are typically recognized as being used for that purpose.

Zivanovic8 replied:
I have changed it to Set.

amyyeung17 commented on core/templates/components/skill-selector/skill-selector.component.ts:
```diff
@@ -199,9 +200,11 @@ export class SkillSelectorComponent implements OnInit {
-      return val.description;
-    });
+    let skills: string[] = this.untriagedSkillSummaries
+      .filter(val => !this.skillIdsToExclude.hasOwnProperty(val.id))
```
Continuing the Set recommendation: we can replace this with just !Set.has(val.id)

Zivanovic8 replied:
I have changed this to Set as well.

amyyeung17 commented on core/templates/pages/skill-editor-page/editor-tab/skill-prerequisite-skills-editor/skill-prerequisite-skills-editor.component.ts:
```diff
@@ -78,6 +78,12 @@ export class SkillPrerequisiteSkillsEditorComponent implements OnInit {
       this.groupedSkillSummaries.others
     );
     const allowSkillsFromOtherTopics = true;
+    const skillIdsToExclude = {
+      [this.skill.getId()]: true,
+    };
+    this.skill.getPrerequisiteSkillIds().forEach(prerequisiteSkillId => {
+      skillIdsToExclude[prerequisiteSkillId] = true;
+    });
```
Continuing the Set recommendation: We can create the set by just passing the entire array plus this.skill.getId(). ES6 allows us to do this in a single line: new Set([...this.skill.getPrerequisiteSkillIds(), this.skill.getId()]).

Zivanovic8 replied:
I have changed this to Set as well.

amyyeung17 commented on core/templates/pages/topics-and-skills-dashboard-page/skills-list/skills-list.component.ts:
```diff
@@ -274,6 +274,7 @@ export class SkillsListComponent {
       this.skillsCategorizedByTopics;
     let untriagedSkillSummaries: SkillSummary[] = this.untriagedSkillSummaries;
     let allowSkillsFromOtherTopics: boolean = true;
+    const skillIdsToExclude = {};
```
Since we don't want any skills to be excluded here, could we make that the default case within the skill-selector component. This way, we wouldn't need to pass an empty object each time.

Zivanovic8 replied:
I have removed this part, because I thought that it was not necessary.

Helper2020 committed:
Modifed test in general_purpose_linter_test (#22063)

* Modifed test in general_purpose_linter_test

* Added comments

* Made suggested changes

Mayank77maruti committed:
Fix #21077 : "Error -11" raised in backend tests.  (#21618)

* test for error -11 added

* test for error -11 added

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils_test.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Update concurrent_task_utils.py

* Updated according to suggestions

Signed-off-by: Mayank Mohapatra <125661248+Mayank77maruti@users.noreply.github.com>

* tests added and updates

* minor updates inplemented

Signed-off-by: Mayank Mohapatra <125661248+Mayank77maruti@users.noreply.github.com>

* Update concurrent_task_utils_test.py

* Update run_backend_tests.py

---------

Signed-off-by: Mayank Mohapatra <125661248+Mayank77maruti@users.noreply.github.com>
Co-authored-by: Sean Lip <sean@seanlip.org>

himkar-singh committed:
Fix #21745: Back button and navbar menu button overlap on small mobile screens. (#21850)

* Fix navbar back button overlapping on small screens

* Fix failing test

* Fix Lint check

* Fix failing lint check

* Update top-navigation-bar.component.ts

* Update top-navigation-bar.component.ts

* updated shouldHideBackButton

* Update top-navigation-bar.component.html

* Update top-navigation-bar.component.ts

* use prettier

* update logic

* use prettier

* use preetier

* updated to use Location

* Fix Failing tests

* Fix Failing tests

* Fix Failing test

* resolve merge conflict

* Update top-navigation-bar.component.ts

* Update top-navigation-bar.component.ts

* Updated the code

* Fix Failing test

* Updated logic

* Additional changes for Fix-BlogPage-back-button

* Update top-navigation-bar.component.ts

* updated the name

* fix lint

* Fix lint check

* Fix failing test

* Updated the function

* Added test cases

* Change the name of test case

* change test name

---------

Co-authored-by: Himkar <himkarsingh00@gmail.com>

himkar-singh committed:
Fix #20197 404 Not Found error on valid collection player page. (#22081)

* Fix 404 error

* Fix 404 errors

galaxyhub9 committed:
Fix #22085: Resolved Media Upload  while Creating a New Skill (#22086)

can access the media section while creating a new skill

masterboy376 committed:
Fix part of #20124: Adds logger beam jobs for CD admin stats generation (#21960)

* additional test case

* added separate job to return logs

* changed logs

* new log job

* fixed coverage issue

Mayank77maruti committed:
Fix #21744 : [CI Failure]: Unhandled error: HttpError: Resource not accessible by integration In Full stack test (#21831)

* Update full_stack_tests.yml

* Update full_stack_tests.yml

* Create full_stack_test_comment.yml

* Delete .github/workflows/full_stack_test_comment.yml

* Create comment_on_ci_workflow.yml

* Update comment_on_ci_workflow.yml

* Update backend_unit_tests.yml

* Update frontend_unit_tests.yml

* Update backend_unit_tests.yml

* Update comment_on_ci_workflow.yml

* Update comment_on_ci_workflow.yml

* Update comment_on_ci_workflow.yml

* Update comment_on_ci_workflow.yml

* Update comment_on_ci_workflow.yml

* Update comment_on_ci_workflow.yml

* Update .github/workflows/comment_on_ci_workflow.yml

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

---------

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

Somnath-Mishra committed:
fix #19636: Beam job change user null bio to empty string in UserSettings model (#21772)

* fix #19636: Change user_bio null to empty string using beam job

* add test user with null bio, short bio, long bio

* update assert statement message in test file

* remove unwanted packages

* fix linter and mypy type checks

* remove unwanted result printing in output section

* correct test function name

* fix linter extra new line at the end of file

HardikGoyal2003 committed:
Fix part of #22082: Migrate angular.copy() to cloneDeep (#22067)

* Removed instances of

* Removed unused function getLastAnswerOnDisplayedCard

* Removed the instances of

* Migrated the base-validation test file

* Removed instances of angular.copy

* Added cloneDeep

* lint fix

* fix frontend tests

* lint fix

ProfessionalCaddie committed:
Fix part of #21214, replace imghdr.what() with filetype.guess() (#21555)

* changed imghdr.what instances into filetype.guess instances and changed utils_test.py by one word and updated requirements files

* changed error message in resources_test to match one from filetype

* changed ACCEPTED_IMAGE_FORMATS_AND_EXTENSIONS slightly to accomodate filetype module and changed variable names and linting per reviews

* changed feconf.ACCEPTED_IMAGES_AND_FORMATS to be in line with filtype module rather than imghdr

* changed feconf lines 417-420 back to original constants

* changed validate_image_and_filename. Split conditional that verifies image into 2. Also further defined test in resources_test.py by replacing upload_bad_image with test_upload_courrupted_image and test_upload_unsupported_image_format

* added tests in image_validation_serviice_tests

* changed order of tests in image_validation_services_test so that a file with unsupported image format is tested earlier

* changed post_json() inside test_upload_image_format to act as if passing png but really passing bmp data

* file_details might be None if the file does not match any of the types that filetype supports.

* removed filetype from requirements.in and regenerated requirements.txt

* Revert "removed filetype from requirements.in and regenerated requirements.txt"

This reverts commit 5a37dc05f6cdcbfdb41fd01127080e66e7e00908.

Ashu463 committed:
Fix #21294 : Drop inactive contributors from issues that they aren't working on, Iteration#2 (#21568)

* changes intiated -> workflow, script, and makefile

* script updated => correct method imported

* testing in progress

* all tests are successful

* removed my branch from workflow

* test cases added for inactive issue checker function

* updated requirement_dev file while liniting

* updated test cases to fix lint issue

* adding types to params and functions in script and test file

* changes to make mypy check successful

* Lint checks are now fine with CI workflow file

* completed all lint checks except for just one

* tried something else to get rid of import error

* recommended changes done

* all lint checks are passing now

* lint passed over main script

* passed lint checks but stuck in mypy type checks

* fixed mypy type check error

* all tests are working fine

* hoping everything will be fine now

* added types to test function

* PyGithub added over requirements_dev.txt

* Pygithub added to requirements_dev

* hash of PyGithub added

* replaced PyGithub with requests

* lint passed on CI workflow

* test file added to shards.json

* permissions: read-all added to workflow

* more test cases added

* test case added

* #pragma added to main function

* test fixed

* fix: iteration#1 while review

* fix: lint fixed on iteration#1

* fix: lint fixed on iteration#1

* fix: CI workflow upgraded on iteration#3

* fix: make command removed

* chore: testing out workflow without installation of requests

* chore: requests library added to requirements_dev.txt

* chore: branch name corrected

* chore: added requests dependency on requirements_dev.in file

* fix: iteration#2 suggestion

* fix: iteration#2 lint and test cases updated

* fix: iteration#4 suggested by sean

* fix: iteration#4 as per sean recommendation

* fix: TypedDict class changed to basic python object

* fix: typos fixed on iteration#5

* fix: iteration#5 linked issue will be picked from dev section of PR page

* fix: well structured objects are made over whole script

* fix: lint errors on script

* fix: test script updated according to new script and patchers added

* fix: lint of test script updated

* fix: cron job schedule updated

* feat: updated issue script to send comment on 7 days and unassign on 10 days

* fix: updated test script

* fix: iteration#6 main script

* fix: test cases written for updated script

* fix: Reverted changes of image_service file

* fix: test script and lint checks

* testing requests over requirment file

* fix: removed requests from requirement_dev.in

* fix: hash removed from requirements_dev.txt

* chore: added current branch over CI workflow

* fix: typos suggestions

* fix: updated test script, and requirements file

* fix: permission set to read-all

* fix: test case covered to 100%

* fix: lint errors

* fix: trying to fix requests issue

* fix: adding requirements.txt in cache path

* fix: added prod path to CI runner

* fix: solved requests error and removed unnecessary things

* fix: replaced hard coded path with pwd cmd

* fix: making this PR to merge.

* feat: trying out another method to resolve requests issue.

* fix: fixed things

* fix: added install dependency to main function

* fix: push_jobs removed from workflow

* fix: test cases for main function

* fix: alert and unassgin comment

* fix: added \n on test case

* fix: added \n to a para

* fix: typecasted INACTIVE DAYS to int at cmnts

* fix: revoked testing credentials

* fix: 2023 -> 2025

---------

Co-authored-by: Sean Lip <sean@seanlip.org>

ojasvi004 committed:
Fix #19975: Ensure misconception feedback option state is saved correctly (#22032)

* resolve bug with misconception feedback checkbox not saving state

* refactor question misconception editor to manage feedback and update test

* refactor question misconception editor to manage feedback and update test

* refactor question misconception editor to manage feedback and update test

* update test

* update test

* refactor question misconception editor to manage feedback and update test

* refactor question misconception editor to manage feedback and update test

omgupta-iitk committed:
Fix part of #21646: Added CUJ 4.1 - See statistics and previous explorations on the creator dashboard (#22042)

* Fix part of #21646: Added CUJ 4.1

* Resolve comments and fixed test-duration

* fix: removed flag from save exploration function

Harish-1511 committed:
Fix 16735 remove window gifframes (#22037)

* Added utility module for ANSI text colorization

* Integrated colorization utility into CI acceptance tests for enhanced readability.

* Integrated colorization utility into CI acceptance tests for enhanced readability.

* Integrated colorization utility into CI acceptance tests for enhanced readability.

* Integrated colorization utility into CI acceptance tests for enhanced readability.

* Integrated colorization utility into CI acceptance tests for enhanced readability.

* Add backend test file for color_utils.py

* Fix linting issues in color_utils_test.py

* Fix linting issues in color_utils_test.py

* Fix linting issues in color_utils.py: added docstrings, fixed formatting, and resolved import

* Fix linting issues in color_utils.py: added docstrings, fixed formatting, and resolved import

* Fix linting issues in color_utils.py: added docstrings, fixed formatting, and resolved import

* Fix linting issues in color_utils.py: added docstrings, fixed formatting, and resolved import

* Fix linting issues in color_utils.py: added docstrings, fixed formatting, and resolved import

* Fix linting issues in color_utils.py: added docstrings, fixed formatting, and resolved import

* Fix linting issues in color_utils.py: added docstrings, fixed formatting, and resolved import

* Fix linting issues in color_utils.py: added docstrings, fixed formatting, and resolved import

* Fix linting issues in color_utils.py: added docstrings, fixed formatting, and resolved import

* Fix linting issues in color_utils.py: added docstrings, fixed formatting, and resolved import

* Fix linting issues in color_utils.py: added docstrings, fixed formatting, and resolved import

* Add color_utils_test to backend test shards

* aligment

* Removed the global window.GifFrames assignment and updated the code to use a direct gif-frames import

* Fix comment style issues to resolve lint check errors

* Fix format in  issues to resolve lint check errors

* Add gif-frames.d.ts to declare 'gif-frames' module and resolve TS7016 error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Replace spyOn(gifFrames, 'default') with a direct jasmine.createSpy override on gifFrames to return a resolved promise in tests.

* Create a mutable alias for gifFrames in tests to allow spying without read-only errors.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Update gif-frames.import.ts to use a default export and modify tests to spy on gifFrames directly, fixing the missing default() method error.

* Fixing the property correction in the processGIFImage of the image-editor.component.ts file.

* Fixing the property correction in the processGIFImage of the image-editor.component.ts file.

* Cast frameData to any[] in processGIFImage to resolve error when accessing its length property

* Replace 'any[]' cast with explicit inline type for frameData.

* Fixing the mypy test occured  errors.

* Fixing the mypy test occured  errors.

* Fixing the mypy test occured  errors.

* Fixing the unknown in the getFrames return type to Promise

* Fixing the unknown in the getFrames return type to Promise

* Fixing the lint errors

* Fixing the lint errors

* Fixing the error occured in the frontend tests

* Add GifFramesService to providers.

* Removing unused written materials.

* Fix dependency injection issue by adding @Injectable({ providedIn: 'root' }) to GifFramesService.

* Remove global GifFrames comments.

* Fixing the lint check.

* Adjusting gif-frames comments.

D-z-V committed:
Fix #14840: Preserve Space after Punctuation (#22114)

* fix: preserve space after punctuation

* fix: output roughly equals test

mosin74 committed:
Fix # 20873 :  Add GA tracking for 'Donate' , 'Volunteer' links in the footer (#22117)

Implemented test and logic for binding event on plain HTML  for footer donate volunteer

Mayank77maruti committed:
Fix part of #21646: Added CUJ 3.2. Access dashboards and other pages from the profile menu (#22106)

* acceptance test for accessing dashboards from profile menu

Signed-off-by: Mayank Mohapatra <125661248+Mayank77maruti@users.noreply.github.com>

* Update access-dashboards-from-profile-menu.spec.ts

* Update access-dashboards-from-profile-menu.spec.ts

* Update logged-in-user.ts

* Update logged-in-user.ts

* Update access-dashboards-from-profile-menu.spec.ts

* Update access-dashboards-from-profile-menu.spec.ts

---------

Signed-off-by: Mayank Mohapatra <125661248+Mayank77maruti@users.noreply.github.com>

seanlip committed:
Introduce GitHub Discussions template for debugging docs (#22132)

Introduce Discussion template for debugging docs

Shriyam-Avasthi committed:
Fix #18390: Completely cover "rte-helper-modal.controller.ts" with unit tests (#21995)

* Fully cover rte helper modal with unit tests

* Update check_frontend_test_coverage.py to address merge conflicts properly

KartikSuryavanshi committed:
Added acceptance test for 4.3 - Create an exploration with multiple states,answer groups, hints and solutions. (#22137)

* Added acceptance test for Create an exploration with multiple states, answer groups, hints and solutions.

* fixed lint checks

* fixed lint check failure

* Update acceptance.json

* Update acceptance.json

* Update create-exploration-with-multiple-states-answer-group-hint-solution.spec.ts

* Made changes requested

* fixed lint checks

Hemant2A2 committed:
Fixes part of #21646: Add CUJ 1.16 View and search for blog posts (#22105)

* added cuj view and search for blog posts

* fix failing checks

* fix acceptance test

* fix tests

* fix tests

* minor correction

* add changes

* fix lint errors

* store all html constants

Zivanovic8 commented:
Hey @amyyeung17,
Apologies for the delayed response, I was on vacation all week. I'll be getting back to the issue right away and will make sure to post some updates by tomorrow. Thanks for your patience!

HardikGoyal2003 committed:
Fix part of #22082: Migrate ng-if to *ng-if (#22118)

* Removed instances of ng-if

* Removed instances of ng-hide

* Removed instances of ng-show

* Removed instances of ng-click

* Removed instances of

* Removed instances of

* Removed instances of

* Removed instances of ng-model

* Removed instances of unit-test-utils.ajs.ts

* lint fix

* lint fix

* fix frontend tests

mon4our committed:
Fix #22096: Fixed hint modal not displaying issue (#22107)

Fixed issue

Nik-09 committed:
Removes queue yaml from the codebase (#22157)

* Removes queue yaml from the codebase

* Adds code comment

omgupta-iitk committed:
Fix #21633: Added screenshot feature for failed acceptance tests (#22043)

* Acceptance Tests, Screenshot feature added

* Added the custom-jest-environment file to root files config

* Fixed Screenshot capturing on the retry attempt

* Fixed comments

* Fixed lints

* Resolve comments: added failure terms

* resolve comment: fixed the function name

* fix path in test-constant file

* Resolve comments: removed close browser object

* fix CI failure

SD-13 committed:
Fix stale review notifier github action (#22189)

Update hash with discussion category fix

Nik-09 committed:
Automatic voiceover regeneration for exploration contents (#21732)

* Creats helper modules.

* Adds platform file for Azure TTS.

* Removes unwanted files

* Adds backend tests

* Add functionality to regenerate voiceovers from exploration editor

* Updates the branch with latest changes

* Fixes backend tests

* Updates docstring

* Fixes backend tests

* Fixes mypy checks and lint checks

* Fixes lint issues

* Adds backend test files in a shard.

* Updates dummy voiceovers!

* Adds backend coverage tests

* updates typo + shards

* temp

* remove comma from backend shard

* Adds saving functionality for generated voiceovers.

* Updates method names and removes consoles

* updates graph coloring after voiceover regeneration

* Adds backend unit tests

* Adds frontend tests

* Adds frontend test coverage

* Fixes mypy checks

* Fixes ts checks

* Hides feature behind feature flag.

* Fixes mypy checks

* Adds backend coverage tests

* Fixes acceptance test

* Addresses review comments

* adds input variable

SD-13 committed:
Update review notifier hash to debug discussion deletion (#22208)

debugging discussion deletion

himkar-singh committed:
Fix #18535:Drop flags for checkpoint_celebration and end_chapter_celebration (#22182)

* remove checkpoint and end chapter celebration flag

* Fix failing test

* Fix test cases

* Remove other test cases

Zivanovic8 committed:
Changed to use Set for skill membership check

Mayank77maruti committed:
Fix #18469: Disable the Submit button for interactions when the input is not valid (#22120)

* disable submit button for invalid input

Signed-off-by: Mayank Mohapatra <125661248+Mayank77maruti@users.noreply.github.com>

* disable submit button and show error when submit button is clicked

* disable submit button and show error when submit button is clicked

* disable submit button and show error when submit button is clicked

* disable submit button and show error when submit button is clicked

* disable submit button test

* disable submit button test

* disable submit button test

* Update current-interaction.service.ts

* Update extensions/interactions/RatioExpressionInput/directives/oppia-interactive-ratio-expression-input.component.spec.ts

Co-authored-by: Nikhil <nikhil.agarwal.2019@gmail.com>

* Update oppia-interactive-fraction-input.component.spec.ts

* Update oppia-interactive-fraction-input.component.spec.ts

---------

Signed-off-by: Mayank Mohapatra <125661248+Mayank77maruti@users.noreply.github.com>
Co-authored-by: Nikhil <nikhil.agarwal.2019@gmail.com>

kevintab95 committed:
Update versions to 3.4.5 (#22219)

* Update core/feconf.py

* Update package.json

---------

Co-authored-by: github-actions[bot] <github-actions@noreply.com>

Mayank77maruti committed:
Fix part of #21643: Workflow To Assign Reviewers on PTAL Comment (#21705)

* Create assign_reviewer_on_ptal.yml

* Update assign_reviewer_on_ptal.yml

* Update assign_reviewer_on_ptal.yml

* Update .github/workflows/assign_reviewer_on_ptal.yml

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

* Update assign_reviewer_on_ptal.yml

* Update assign_reviewer_on_ptal.yml

* Update assign_reviewer_on_ptal.yml

* Update assign_reviewer_on_ptal.yml

* Update assign_reviewer_on_ptal.yml

---------

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

seanlip committed:
Fix errors in discussion template (#22205)

* Fix errors in discussion template

* Fix more issues

* drop one ID

seanlip committed:
Fix error in proposal review template. (#22236)

masterboy376 committed:
Fixes part of #20124: Fixed the StoryModel does not exists error (#22188)

fixed the stroy model does not exists error

MurarkaT committed:
Fix radio button alignment (#22217)

* Fix radio button alignment

* Radio button issue resolved

seanlip committed:
Add missing label to Discussions template (#22229)

Add missing label

mishrayush15 committed:
Fix #22001, Added reverse iteration in SVG Sanitizer and wrote unit tests for validation (#22135)

* Fix: SVG validation and added unit test for validation

* Fix: Removed unnecessary comments

* Fix: Added meaningful comments

* Fix: Added comments explaing the requirement of reverse iteration

* Fix: Updated comments for SVG Sanitizer Unit Test

* Fix: Improve comments in SVG Sanitizer unit test

FaithAbiola committed:
Fix #16126: Let developers skip installation when running backend tests (#21999)

* skip third party install

* skip third party install

* Refactor imports, rename test functions, and add --skip-install to start.py

* Refactor imports, rename test functions, and add --skip-install to start.py

* Refactor imports, rename test functions, and add --skip-install to start.py

* Refactor imports, rename test functions, and add --skip-install to start.py

* Refactor start script: Move install_third_party_libs call and clean up comments

* Refactor start script: Move install_third_party_libs call and clean up comments

* Refactor start script: Move install_third_party_libs call and clean up comments

seanlip committed:
Move label attribute to the right place in debugging docs discussion template (#22282)

Move label attribute to the right place

ojasvi004 committed:
Fix #21987: Resolve mobile thumbnail modal glitch issue in blog post editor. (#22036)

* resolve mobile UI glitch in blog post thumbnail upload

* resolve mobile UI glitch in blog post thumbnail upload

* resolve mobile UI glitch in blog post thumbnail upload

* improve button UI

* update mobile screenshots for blog editor error message on duplicate titles

* adjust failureTrigger increment in puppeteer utils

* update

* update

Mayank77maruti committed:
Fix part of #21646: Added CUJ 3.9. Edit avatar and view created lessons in the profile page (#22175)

* edit avatar and view exploration from profile menu

Signed-off-by: Mayank Mohapatra <125661248+Mayank77maruti@users.noreply.github.com>

* edit avatar and view exploration from profile menu

Signed-off-by: Mayank Mohapatra <125661248+Mayank77maruti@users.noreply.github.com>

* Update edit-avatar-and-view-exploration-from-profile-menu.spec.ts

* Required changes made

Signed-off-by: Mayank Mohapatra <125661248+Mayank77maruti@users.noreply.github.com>

---------

Signed-off-by: Mayank Mohapatra <125661248+Mayank77maruti@users.noreply.github.com>

Zivanovic8 commented:
Hi @amyyeung17,
Thanks for your recommendation! The fixes worked well, but I’m currently encountering some failures in the full-stack tests. Could you help me identify the issue and suggest how to resolve it? I’d really appreciate any guidance you can provide.

Zivanovic8 committed:
Initialized an empty Set skillIdsToExclude in select-skill-modal component

mon4our committed:
Temporarily Fix #22283: Set @types/babel__traverse to previous stable version (#22300)

* Fixed babel__traverse dependency issue temporarily

* Removed overrides

kevintab95 committed:
Revert "Automatic voiceover regeneration for exploration contents" (#22262)

Revert "Automatic voiceover regeneration for exploration contents (#21732)"

This reverts commit c427e3aaa75f97bfc4e915ed2db9022ea752fbc3.

Zivanovic8 committed:
Merge branch 'develop' into fixSkillPrerequisite

kevintab95 reviewed (approved):
LGTM, thanks @Zivanovic8!

Vir-8 commented on core/templates/components/skill-selector/select-skill-modal.component.ts:
```diff
@@ -42,6 +42,7 @@ export class SelectSkillModalComponent extends ConfirmOrCancelModal {
   selectedSkillId!: string;
   allowSkillsFromOtherTopics: boolean = false;
   associatedSkillSummaries!: ShortSkillSummary[];
+  skillIdsToExclude: Set<string> = new Set<string>();
```
Shouldn't this set be handled such that it could be undefined?

```suggestion
  skillIdsToExclude!: Set<string> = new Set<string>();
```

Vir-8 commented on core/templates/components/skill-selector/skill-selector.component.ts:
```diff
@@ -47,6 +47,7 @@ export class SkillSelectorComponent implements OnInit {
   @Input() categorizedSkills!: CategorizedSkills;
   @Input() untriagedSkillSummaries!: SkillSummary[];
   @Input() allowSkillsFromOtherTopics!: boolean;
+  @Input() skillIdsToExclude: Set<string> = new Set();
```
Same as above

Vir-8 reviewed (commented):
Hi @Zivanovic8, thanks for the PR, one comment.

Vir-8 reviewed (approved):
Misread, my bad, LGTM! @Zivanovic8

seanlip commented:
This PR LGTM. Thanks @Zivanovic8!

Zivanovic8 committed:
Merge branch 'develop' into fixSkillPrerequisite

Zivanovic8 commented:
Hey @seanlip , I was wondering if you know why some of these tests are failing. I've noticed that at different times, certain tests passed while others failed, and now some that previously passed are failing again. Any ideas on what might be causing this?

seanlip commented:
Plse see the instructions at https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR, particularly point 1 in the introduction and the last section of the page. This will tell you what to do in cases like this.

Thanks!

Zivanovic8 committed:
Merge branch 'develop' into fixSkillPrerequisite

Zivanovic8 committed:
Merge branch 'develop' into fixSkillPrerequisite

Zivanovic8 committed:
Merge branch 'develop' into fixSkillPrerequisite

Zivanovic8 commented:
@seanlip all checks passed, I think it is ready for merging.
Thank you!

seanlip commented:
Thanks @Zivanovic8 for fixing this issue!
